odoo/addons/base/tests/test_sql.py

```diff
@@ -41,6 +43,14 @@ def test_sql_with_wrong_pattern(self):
         with self.assertRaises(KeyError):
             SQL("SELECT id FROM table WHERE foo=%(one)s AND bar=%(two)s", one=1, to=2)
 
+    def test_escape_percent(self):
+        sql = SQL("'%%' || %s", 'a')
+        self.assertEqual(sql.code, "'%%' || %s")
+        with self.assertRaises(TypeError):
+            SQL("'%'")  # unescaped percent
+        with self.assertRaises(TypeError):
+            SQL("'%%' || %s")  # missing parameter
```

Comment:
I'd suggest we extend the tests to add cases were we use a combination of string and SQL parameters, and also tests the kwargs version.
For inspiration: https://github.com/odoo/odoo/commit/3cae03ff9e1f5f25993228718b5fb465db6a0442#diff-f98923f47c922db93401c59acd726fca6b638eba6d93003c10a4502bc99ddfdeR197-R205

Reply:
These are all covered by the base examples here.
It would just additional examples of usage.

Reply:
> These are all covered by the base examples here.

I don't see any example where we pass an SQL object to as argument while each use `%`. Nor one where the kwargs variant is used. Taking into account that formatting happens in multiple places depending on whether we pass a kwargs, or if the passed argument is an SQL object or not, I think it's worth having such cases.

Reply:
```py
self.assertEqual(SQL("'foo%%'").code, "'foo%%'")
self.assertEqual(SQL("'foo%%' || %s", 'bar').code, "'%%' || %s")
self.assertEqual(SQL("'foo%%' || %(bar)s", bar='bar').code, "'%%' || %s")
```

Reply:
... and given what we do here 
https://github.com/odoo/odoo/blob/b75e2d323f28c7b8137932d0804a8fb3c32be2de/odoo/tools/sql.py#L113-L116

I believe that we should have that part of the code tested specifically for the `%` issue.
```py
self.assertEqual(SQL("%(foo) AND bar='baz%%'", foo=SQL("qrux='%%'")).code, ...)
self.assertEqual(SQL("%(foo) AND bar='baz%%'", foo=SQL("%s='%%s'", "qrux")).code, ...)
```

Reply:
Tests added.